src/transform/src/analysis/equivalences.rs

```diff
@@ -399,17 +399,29 @@ impl EquivalenceClasses {
         //    TODO: If all lists are sorted, this could be a linear merge among all.
         //          They stop being sorted as soon as we make any modification, though.
         //          But, it would be a fast rejection when faced with lots of data.
-        for index1 in 0..self.classes.len() {
-            for index2 in 0..index1 {
-                if self.classes[index1]
-                    .iter()
-                    .any(|x| self.classes[index2].iter().any(|y| x == y))
-                {
-                    let prior = std::mem::take(&mut self.classes[index2]);
-                    self.classes[index1].extend(prior);
-                    stable = false;
+        //    `expr_to_class_index` tells us for each expression the class index where we last saw
+        //    it.
+        //    `to_merge` has pairs of classes to be merged. The first element of the pair should be
+        //    an earlier class than the second, and `to_merge` should be in sorted order. (These
+        //    invariants are important when handling expressions that appear in more than two
+        //    classes: we'll first merge the first two into the second, and then all this into the
+        //    third, and so on.)
+        let mut expr_to_class_index = BTreeMap::new();
+        let mut to_merge = Vec::new();
+        for (index, class) in self.classes.iter().enumerate() {
+            for expr in class {
+                if let Some(other_index) = expr_to_class_index.get(expr) {
+                    to_merge.push((*other_index, index));
                 }
```

Comment:
I think it's safe to fuse these two loops, with the insert happening after logging the merge: we look up first and then record that we're merging.

(We could even move the insert into the `else` of the `let if` were we to extend `self.classes[ind1]` instead of `self.classes[ind2]`.)

Reply:
With the fused loops, I'd be worried about the situation where the same class has an expression twice, in which case the fused loops would record a merge of a class with itself.

Reply:
Got it. Not a huge deal.

At that rate, shouldn't `to_merge` be a set? Maybe not worth changing, since we'd be saving on calls to `extend` that don't do anything anyway. Using a union-find obviates the need for thinking carefully about this. 😁

Reply:
Hmm, could be a set, but then one more brittleness is that we'd be relying on the ordering of the set. I'd rather keep this simple for now, and then if we find the time to do Union-Find then all this will become more elegant.